gdal/frmts/wms/minidriver_mrf.cpp

```diff
@@ -265,9 +265,9 @@ size_t WMSMiniDriver_MRF::GetIndexAddress(const GDALWMSTiledImageRequestInfo &ti
     // Bottom level is 0
     int l = - tiri.m_level;
     if (l < 0 || l >= static_cast<int>(offsets.size()))
-        return ~static_cast<size_t>(0); // Indexing error
+        return ~0; // Indexing error
```

Comment:
This change and the next one look wrong. ~0 would lead to 0xFFFFFFFFU instead of 0xFFFFFFFFFFFFFFFFU for 64 bit size_t

Reply:
The type of the integer literal will be promoted by the compiler to match the return type of the function. Here is an small progam and GDB session output to illustrate:

    #include <stdlib.h>
    #include <stdio.h>

    size_t myfunc() { return ~0; }

    int main()
    {
      size_t rc = myfunc();
      return 0;
    }

In GDB:

    (gdb) start
    Temporary breakpoint 1 at 0x4004eb: file foo.c, line 8.
    Starting program: /home/bje/foo.x 
    Temporary breakpoint 1, main () at foo.c:8
    8	  size_t rc = myfunc();
    (gdb) step
    myfunc () at foo.c:4
    4	size_t myfunc() { return ~0; }
    (gdb) finish
    Run till exit from #0  myfunc () at foo.c:4
    0x00000000004004f5 in main () at foo.c:8
    8	  size_t rc = myfunc();
    Value returned is $1 = 18446744073709551615
    (gdb) print/x $1
    $2 = 0xffffffffffffffff

Reply:
OK, my bad. Actually ~0 is first evaluated as (int)-1, and (int)-1 is promoted to (size_t)0xFFFFFFFFFFFFFFFFU due to negative signed to unsigned promotion rules.

How did you detect those instances ? Some clang tooling ?

Reply:
I spotted them while reading source code, then started grepping for `static_cast<.*>([0-9]+)` -- one of the wonderful advantages of `static_cast` in C++. Also, GCC has an option `-Wuseless-casts`, but it is not activated by `-Wall`.